digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala

```diff
@@ -205,28 +206,50 @@ trait TrackCalculatorStrategy {
       }
   }
 
-  protected def setCalibrationPoint(pl: ProjectLink, userCalibrationPoint: Option[UserDefinedCalibrationPoint], startCP: Boolean, endCP: Boolean, source: CalibrationPointSource = UnknownSource) = {
+  protected def setCalibrationPoint(pl: ProjectLink, userCalibrationPoint: Option[UserDefinedCalibrationPoint],
+                                    startCP: Boolean, endCP: Boolean, source: CalibrationPointSource = UnknownSource): ProjectLink = {
     val sCP = if (startCP) CalibrationPointsUtils.makeStartCP(pl) else None
     val eCP = if (endCP) CalibrationPointsUtils.makeEndCP(pl, userCalibrationPoint) else None
     pl.copy(calibrationPoints = CalibrationPointsUtils.toProjectLinkCalibrationPointsWithSourceInfo((sCP, eCP), source))
   }
 
-  protected def getUntilNearestAddress(seq: Seq[ProjectLink], endProjectLink: ProjectLink): (Seq[ProjectLink], Seq[ProjectLink]) = {
-    if (endProjectLink.discontinuity == MinorDiscontinuity || endProjectLink.discontinuity == Discontinuous) {
-      val continuousProjectLinks = seq.takeWhile(pl => pl.startAddrMValue < endProjectLink.endAddrMValue)
-
-      if (continuousProjectLinks.isEmpty)
+  /**
+    * Returns project links for the other track before and after the point where there is discontinuity on the track.
+    *
+    * @param trackA
+    * @param linkOnTrackB
+    * @return (Project links before the discontinuity point, project links after the discontinuity point)
+    */
+  protected def getUntilNearestAddress(trackA: Seq[ProjectLink], linkOnTrackB: ProjectLink): (Seq[ProjectLink], Seq[ProjectLink]) = {
+    if (linkOnTrackB.discontinuity == MinorDiscontinuity || linkOnTrackB.discontinuity == Discontinuous) {
+
+      // Sort links in order by the distance from the end of the link on track b.
+      val linkOnTrackBEndPoint = linkOnTrackB.lastPoint
+      val nearestLinks: Seq[(ProjectLink, Double)] = trackA.map(pl => (pl,
+        pl.lastPoint.distance2DTo(linkOnTrackBEndPoint)
+      )).sortWith(_._2 < _._2) // Links with nearest end points first
+
+      // Choose the nearest link that has similar enough angle to the linkOnTrackB and set it as lastLink
+      val linkOnTrackBDirection = linkOnTrackB.lastSegmentDirection
+      val maxAngleBetweenLinks = math.toRadians(67.5) // TODO
```

Comment:
Please do address the magic number 67.5 and the //TODO

Reply:
It's now 75 degrees and there is a comment trying to explain it a bit.